### DISCUSSION TITLE
Trino throwing error when key_column data type is UUID

### DISCUSSION THREAD

**Discussion by Ryuk-me** [erezsh/reladiff#23] [Q&A] [answered]
Hey @erezsh, hope you are doing well.

I'm trying to find difference between trino DB's which is connected to a postgres database
I'm getting this error 
```
raise NotImplementedError(f"Cannot use a column of type {kt} as a key")
NotImplementedError: Cannot use a column of type UnknownColType(text='uuid') as a key
```

Here is my code
```py
from reladiff import connect_to_table, diff_tables

table1 = connect_to_table("trino://admin@localhost/source/public", "raw", "custid")
table2 = connect_to_table("trino://admin@localhost/target/public", "raw", "custid")

for sign, row in diff_tables(table1, table2):
    print(sign, row)
```

Postgres Database rows

![image](https://github.com/user-attachments/assets/fe2d6d7f-b431-4232-b84f-35bf42ef1aae)

Is there any fix?
When I make uuid as TEXT it works fine.

Thank you for your great work.

**Comment by erezsh** [comment]
Hello @Ryuk-me ,

Yes, it looks like Trino doesn't currently support UUID columns.

I think it's relatively easy to add, but it needs to be added to the `sqeleton` package.

If you'd like to open a PR, I can explain to you how to fix it yourself. Otherwise, I'll try to get to it myself in the coming days.

**Reply by Ryuk-me** [reply]
Sure, I'll raise a PR.

**Reply by erezsh** [reply]
@Ryuk-me  Great! 

All you need to do is add the "uuid" type to the trino driver.

- You can see how it's done for postgresql: https://github.com/erezsh/sqeleton/blob/b4476ceed203cfb2a573d09b1fbecbde0b224269/sqeleton/databases/postgresql.py#L82

- Trino is implemented as a subset of presto, so it's probably best to add it to presto here (unless they behave differently for this): https://github.com/erezsh/sqeleton/blob/b4476ceed203cfb2a573d09b1fbecbde0b224269/sqeleton/databases/presto.py#L83

P.S. First thing is to do is set up the env. So clone and install both reladiff and sqeleton for development. (like `pip install -e .`), and make sure the cloned sqeleton is being used.

**Reply by erezsh** [reply]
@Ryuk-me Let me know if you need any help or advice!

**Reply by Ryuk-me** [reply]
Hey @erezsh, thanks for the instructions after making the suggested changes, I'm getting this error
```
raise self._process_error(response["error"], response.get("id"))

trino.exceptions.TrinoUserError: TrinoUserError(type=USER_ERROR, name=FUNCTION_NOT_FOUND, message="Unexpected parameters (uuid) for function trim. Expected: trim(char(x)), trim(char(x), codepoints), trim(varchar(x)), trim(varchar(x), codepoints)", query_id=20240813_123832_00024_3mat3)     
```

Table structure `postgres`

```sql
CREATE TABLE test_table (
    id UUID PRIMARY KEY,
    name VARCHAR(100),
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

INSERT INTO test_table (id, name) VALUES
('550e8400-e29b-41d4-a716-446655440000', 'Nk'),
('3b241101-e2bb-4255-8caf-4136caaae2e6', 'Bob');
```
Changes:
![image](https://github.com/user-attachments/assets/2938669e-d44f-4cb7-89e2-54e361c474f4)

![image](https://github.com/user-attachments/assets/5f4822d5-d1e7-4bf6-920b-92e254b1c8b4)

```py
from reladiff import connect_to_table, diff_tables

table1 = connect_to_table("trino://admin@localhost/source/public", "test_table", "id")
table2 = connect_to_table("trino://admin@localhost/target/public", "test_table", "id")

for sign, row in diff_tables(table1, table2):
    print(sign, row)
```
I have installed `sqeleton` and `reladiff` locally and reinstalled using `pip install -e .` after making changes, as you can see in the below pictures both are installed locally, and both of the clones are UpToDate.

![image](https://github.com/user-attachments/assets/97a9ba16-b30a-4ba9-aaa0-54c242b3c945)
![image](https://github.com/user-attachments/assets/41cd8a02-545f-42b1-8417-59c2fef4451f)

**Reply by erezsh** [reply]
Hi @Ryuk-me , sounds like you made progress!

I think this error is happening because the trino implementation (in sqeleton) of `normalize_uuid()` isn't correct. See, until now, Trino only supported String_UUID, but now that it can also be Native_UUID, this use-case needs to be addressed.

- The default implementation either converts to char or not, based on the type: https://github.com/erezsh/sqeleton/blob/master/sqeleton/abcs/mixins.py#L46

- Mysql takes the "easy" way and always casts to char: https://github.com/erezsh/sqeleton/blob/master/sqeleton/databases/mysql.py#L48

Hopefully that moves you forward! If not, create a draft pull-request with your changes. It will be easier to debug if I could run it myself.

**Reply by Ryuk-me** [reply]
Hey @erezsh, I've raised a draft PR please have a look whenever you're free.

PR 
https://github.com/erezsh/sqeleton/pull/20

**Reply by erezsh** [reply, marked as answer]
@Ryuk-me I created a PR that is supposed to solved this. Can you give it a try and let me know if it works for you too?  

https://github.com/erezsh/reladiff/pull/28

**Reply by Ryuk-me** [reply]
Hey @erezsh, its working now.
Great work!! 
Thank you.
